### PR TITLE
10 address structured fp

### PR DESCRIPTION
**Pull request by fpodoo**
Proposition of module to have a more structured address:
- street ==>  Street + Street Number + ' - ' + Door Number
- city     ==> city_id

This module is required for some localisations like Mexico. (legal reports need these extra fields)
## 

I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by gdgellatly** [comment]
Personally I think the re.compile pattern at the start is best done as a
field in language localization or similar.

As an aside, compiling regex explicitly doesn't have much benefit anymore,
as python implicitly compiles and caches. But it does make code look
cleaner IMO.

**Comment by lukebranch** [comment]
I was thinking about building something to handle the large number of different country-specific standards for address formatting previously, however it would be incredibly useful to have this as an optional module in the core.

Here's a useful guide to all the different countries and their accepted address format:

https://www.informatica.com/hk/products/data-quality/data-as-a-service/address-verification/address-formats.html

I would be happy to split this into a complete set of necessary fields and then list them out by country and layout for localisation if it would help.

This would ensure far more accurate postal address formatting for Odoo and avoid returned parcels due to incorrectly formed addresses.

**Comment by nhomar** [comment]
Topics to be solved on this PR or another one:

- Warning of access rules to city model: res.country.state.city
- [Views]: The normal form view has an incorrect order.
selection_018
- [Views]: PopUp View to create partner is not showing the new fields and is not hiding the old city field.
selection_020
- [Views and company]: The company should be consistent.
selection_024


Do we fix this after merge or do we complete it before merge?

**Comment by pedrobaeza** [comment]
One city can have several zip codes. That's why we have in community this module: https://github.com/OCA/partner-contact/tree/9.0/base_location. Please check and see also the modification in other views.

**Comment by pedrobaeza** [comment]
@nhomar, @moylop260 I write here for not continuing hijacking the mail thread.

The module (base_location) is not Spanish. It's used by half of the Odoo world:

* https://github.com/OCA/l10n-spain/tree/8.0/l10n_es_toponyms
* https://github.com/OCA/l10n-france/tree/8.0/l10n_fr_base_location_geonames_import
* https://github.com/OCA/l10n-italy/tree/8.0/l10n_it_base_location_geonames_import
* https://github.com/OCA/l10n-switzerland/tree/8.0/l10n_ch_zip
* ...

and it has a broader scope than your approach. 

@fpodoo, please take this into account:

We should include here also a model for zip entries. That zip entries can have a parent city. Then, you will have two many2one selectors: one for the city and another for the zip. You can directly search for a zip or a city name and select the correct zip entry, and all data (including `zip` field) are filled (the same as is now on _base_location_ module), or you can search the city, and all data except zip are filled, but zip table selector domain is restricted to those that have as parent the selected city.

This way, we can accommodate all the requirements and have a better module for all. 

One doubt I have, @nhomar: AFAIK, there's also counties and other territorial divisions in Mexico that can have also a zip entry. Is that true? If so, well, you can add additional columns in l10n_mx for specifying parent counties and so on.

File: addons/base_address_structured/__init__.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import models
```
Comment (pedrobaeza) [inline addons/base_address_structured/__init__.py]: relative import avoids possible side problems `from . import models`

File: addons/base_address_structured/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+{
+    'name': 'Structured Addresses',
+    'summary': 'Split street in several fields',
+    'sequence': '19',
+    'category': 'Base',
+    'complexity': 'easy',
+    'description': """
+Split Addresses
```
Comment (pedrobaeza) [inline addons/base_address_structured/__manifest__.py]: Better to put this in a README.rst file

File: addons/base_address_structured/models/__init__.py [hunk 3]
```diff
@@ -0,0 +1,4 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import base_address_structured
```
Comment (pedrobaeza) [inline addons/base_address_structured/models/__init__.py]: relative import

File: addons/base_address_structured/models/base_address_structured.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    code = fields.Char('Code')
+
+
+class PartnerAddress(models.Model):
+    _inherit = ['res.partner']
+    _name = 'res.partner'
+
+    street_raw = fields.Char('Street')
+    street_number1 = fields.Char('Street Number')
+    street_number2 = fields.Char('Door Number')
```
Comment (pedrobaeza) [inline addons/base_address_structured/models/base_address_structured.py]: There are lots of other divisions that are needed across the world. For example, here in Spain we have "Building" and "Floor" divisions also, so our address is:

*Street, number, building, floor, door number*

My suggestion here is to stay as is (as a contact name stays neutral without all the first name, last name, middle initial and so on considerations), or to go for an sparse field that allows to store a data structure, and define by res.country the needed "sub-fields".

**Comment by pedrobaeza** [comment]
One possible foundation to handle addresses around the world: https://github.com/openvenues/pypostal

File: addons/base_address_structured/models/base_address_structured.py (hunk 4 above)
Comment (pedrobaeza) [inline addons/base_address_structured/models/base_address_structured.py]: One possible foundation to handle addresses around the world: https://github.com/openvenues/pypostal

File: addons/base_address_structured/__manifest__.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+{
+    'name': 'Structured Addresses',
+    'summary': 'Split street in several fields',
+    'sequence': '19',
+    'category': 'Base',
+    'complexity': 'easy',
```
Comment (dreispt) [inline addons/base_address_structured/__manifest__.py]: Deprecated?

File: addons/base_address_structured/__manifest__.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+For legal reports, some countries need to split the street into several fields,
+with the street name, the house number, and room number. This module also
+replace the city by a relation field.
+        """,
+    'data': [
+        'views/base_address_structured.xml'
+    ],
+    'depends': ['base'],
+}
```
Comment (dreispt) [inline addons/base_address_structured/__manifest__.py]: Missing author

**Comment by eLBati** [comment]
Anyway, I would avoid adding new functionalities to a stable version.

https://github.com/odoo/odoo/wiki/Contributing#what-does-stable-mean :smirk:

**Comment by alexis-via** [comment]
@eLBati It's a new module, and it's not auto-installable, so it's not a problem for a stable version. If you don't want it, you just don't install it.

**Comment by luistorresm** [comment]
@fpodoo 

I found the next error when try generate a new partner without street:

[
new odoo
](https://youtu.be/gm95E4kbMfg)

These new fields street_number1, street_number2 & city_id, could be consider to add also in the company address?

Regards

**Comment by pedrobaeza** [comment]
Any news on this?

**Comment by OpenCode** [comment]
Any news on this?

**Comment by nhomar** [comment]
Still in progress, try the last versions of the modules with changes @qdp-odoo considering some mentioned points.

In the plans there is some refactors, like:

- Use an specific view country dependent.
- Have an special new type of field to manipulate and load cities (pending).
- Cleanup the logic to be cleanest and more inheritable.

Be in touch.

File: odoo/addons/base/res/res_city.py [hunk 7]
```diff
@@ -1,15 +1,16 @@
...
 
 
 class City(models.Model):
     _name = 'res.city'
     _description = 'City'
     _order = 'name'
 
-    name = fields.Char(string='City Name', required=True, translate=True)
-    zipcode = fields.Char(string='Zip')
+    name = fields.Char("Name", required=True, translate=True)
```
Comment (nhomar) [inline odoo/addons/base/res/res_city.py]: AFAICR string is not necessary if field name is the same isn't¿?

File: odoo/addons/base/res/res_country.py [hunk 8]
```diff
@@ -32,30 +32,37 @@ class Country(models.Model):
...
     image = fields.Binary(attachment=True)
     phone_code = fields.Integer(string='Country Calling Code')
     country_group_ids = fields.Many2many('res.country.group', 'res_country_res_country_group_rel',
                          'res_country_id', 'res_country_group_id', string='Country Groups')
     state_ids = fields.One2many('res.country.state', 'country_id', string='States')
-    enforce_cities = fields.Boolean(string='Enforce Cities', help="Check this box to ensure every address created \
-in that country has a value for the field 'City' chosen in the list of the country's cities. Note that users with \
-the 'Partner Manager' group are still able to create new cities for a country.")
+    enforce_cities = fields.Boolean(
+        string='Enforce Cities',
```
Comment (nhomar) [inline odoo/addons/base/res/res_country.py]: Remove string I think is not necessary

File: addons/base_address_extended/models/base_address_extended.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+             "\n%(street_name)s: the name of the street"
+             "\n%(street_number)s: the house number"
+             "\n%(street_number2)s: the door number",
+        default='%(street_number)s/%(street_number2)s %(street_name)s', required=True)
+
+class Partner(models.Model):
+    _inherit = ['res.partner']
+    _name = 'res.partner'
+
+    street_name = fields.Char('Street Name', compute='_split_street',
```
Comment (moylop260) [inline addons/base_address_extended/models/base_address_extended.py]: https://www.odoo.com/documentation/10.0/reference/guidelines.html
```md
Method conventions
Compute Field : the compute method pattern is _compute_<field_name>
```

File: addons/base_address_extended/models/base_address_extended.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+                    #maxsplit set to 1 to unpack only the first element and let the rest untouched
+                    tmp = street_raw.split(separator, 1)
+                    if len(tmp) == 2:
+                        field_value, street_raw = tmp
+                        vals[field_name] = field_value
+                if field_value or not field_name:
+                    # select next field to find (first pass OR field found)
+                    # [2:-2] is used to remove the extra chars '%(' and ')s'
+                    field_name = re_match.group()[2:-2]
+                else:
```
Comment (moylop260) [inline addons/base_address_extended/models/base_address_extended.py]: Is required this `else` sentence?

File: addons/base_address_extended/models/base_address_extended.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+                    # value not found: keep looking for the same field
+                    pass
+                if field_name not in street_fields:
+                    raise UserError(_("Unrecognized field %s in street format.") % field_name)
```
Comment (moylop260) [inline addons/base_address_extended/models/base_address_extended.py]: What about use comma instead `%`?
`raise UserError(_("Unrecognized field %s in street format."), field_name)`

File: addons/base_address_extended/__manifest__.py [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+
+This module holds all extra fields one may need to manage accurately addresses.
+
+For example, in legal reports, some countries need to split the street into several fields,
+with the street name, the house number, and room number.
+        """,
+    'data': [
+        'views/base_address_extended.xml'
+    ],
+    'depends': ['base'],
```
Comment (moylop260) [inline addons/base_address_extended/__manifest__.py]: Is not need it author, license, installable keys?

File: addons/base_address_extended/__init__.py [hunk 13]
```diff
@@ -0,0 +1,4 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import models
```
Comment (moylop260) [inline addons/base_address_extended/__init__.py]: What about `from . import models`?

File: addons/base_address_extended/models/__init__.py [hunk 14]
```diff
@@ -0,0 +1,4 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import base_address_extended
```
Comment (moylop260) [inline addons/base_address_extended/models/__init__.py]: What about `from . import base_address_extended`?

File: addons/base_address_extended/__manifest__.py [hunk 15]
```diff
@@ -0,0 +1,22 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+{
+    'name': 'Extended Addresses',
+    'summary': 'Add extra fields on addresses',
+    'sequence': '19',
+    'category': 'Base',
+    'complexity': 'easy',
+    'description': """
```
Comment (moylop260) [inline addons/base_address_extended/__manifest__.py]: What about use README.rst?

File: addons/base_address_extended/models/base_address_extended.py [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+    @api.depends('street', 'country_id.street_format')
+    def _split_street(self):
+        """Splits street value into sub-fields.
+        Recomputes the fields of STREET_FIELDS when `street` of a partner is updated"""
+        street_fields = self.get_street_fields()
+        for partner in self:
+            if not partner.street:
+                partner.street_name = ''
+                partner.street_number = ''
+                partner.street_number2= ''
```
Comment (moylop260) [inline addons/base_address_extended/models/base_address_extended.py]: Small lint of Ssace before assignment `partner.street_number2 = ''`

File: addons/base_address_extended/models/base_address_extended.py [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+            # last field value is what remains in street_raw minus trailing chars in street_format
+            trailing_chars = street_format[previous_pos:]
+            if trailing_chars and street_raw.endswith(trailing_chars):
+                vals[field_name] = street_raw[:-len(trailing_chars)]
+            else:
+                vals[field_name] = street_raw
+            # assign the values to the fields
+            # /!\ Note that a write(vals) would cause a recursion since it would bypass the cache
+            for k, v in vals.items():
+                partner[k] = v
```
Comment (suniagajose) [inline addons/base_address_extended/models/base_address_extended.py]: `False` values should be avoided

``` python
if v is not False:
    partner[k] = v
```

**Comment by qdp-odoo** [comment]
branch has been merged in master at revisions: 4dd27bba346e50f40769d1412fbca0b5f65bb5a6, 0dc4e13f145a449395aa82cfe75dfbe09afb5455 and a14d6bb9acccc7025439360f4a3250ef23eb94ec

**Pull request closed without merging**